nmagent/cachedclient.go

```diff
@@ -0,0 +1,164 @@
+func (c *CachedClient) SupportedAPIs(ctx context.Context) ([]string, error) {
+	supportedAPIs, err := c.ClientIF.SupportedAPIs(ctx)
+	if err != nil {
+		return supportedAPIs, errors.Wrap(err, "failed to retrieve supportedApis from nmagent")
+	}
+	c.updateNMASupportedApisCache(supportedAPIs)
+	return supportedAPIs, nil
+}
+
+// GetHomeAz returns homeaz cache directly
+func (c *CachedClient) GetHomeAz(_ context.Context) (HomeAzResponse, error) {
+	return c.readHomeAzAndErrorCache()
+}
+
+// isAPISupportedByNMAgent checks cache to see if the given api supported by nmagent client.
+func (c *CachedClient) isAPISupportedByNMAgent(api string) bool {
+	supportedAPIs := c.readNMASupportedApisCache()
+	for _, supportedAPI := range supportedAPIs {
+		if supportedAPI == api {
+			return true
+		}
+	}
+	return false
+}
+
+// readNMASupportedApisCache gets the nmaSupportedApisCache value
+func (c *CachedClient) readNMASupportedApisCache() []string {
+	c.supportedApisMu.RLock()
+	defer c.supportedApisMu.RUnlock()
+	supportedApis := c.nmaSupportedApisCache
+	supportedApisCopy := make([]string, len(supportedApis))
+	copy(supportedApisCopy, supportedApis)
+	return supportedApisCopy
+}
+
+// updateNMASupportedApisCache updates the nmaSupportedApisCache value
+func (c *CachedClient) updateNMASupportedApisCache(supportedApis []string) {
+	c.supportedApisMu.Lock()
+	defer c.supportedApisMu.Unlock()
+	c.nmaSupportedApisCache = supportedApis
+}
+
+// readHomeAzAndErrorCache get the homeAzResponseCache and homeAzResponseErrorCache value
+func (c *CachedClient) readHomeAzAndErrorCache() (HomeAzResponse, error) {
+	c.homeAzMu.RLock()
+	defer c.homeAzMu.RUnlock()
+	homeAzResponse := c.homeAzResponseCache
+	err := c.homeAzResponseErrorCache
+	return homeAzResponse, err
+}
+
+// updateHomeAzAndErrorCache update the homeAzResponseCache and homeAzResponseErrorCache value
+func (c *CachedClient) updateHomeAzAndErrorCache(homeAzResponse HomeAzResponse, err error) {
+	c.homeAzMu.Lock()
+	defer c.homeAzMu.Unlock()
+	c.homeAzResponseCache = homeAzResponse
+	c.homeAzResponseErrorCache = err
+}
+
+// Start starts a new thread to refresh home az cache
+func (c *CachedClient) Start(retryIntervalInSecs int) {
+	go c.refresh(retryIntervalInSecs)
+}
+
+// Stop ends the refresh thread
+func (c *CachedClient) Stop() {
+	close(c.closing)
+}
+
+// refresh keeps retrying until successfully getting home az from nmagent
+func (c *CachedClient) refresh(retryIntervalInSecs int) {
+	for {
+		select {
+		case <-c.closing:
+			return
+		default:
+			ctx, cancel := context.WithTimeout(context.Background(), ContextTimeOut)
+			homeAzResponse, populateErr := c.populateHomeAzCache(ctx)
+			// keep retrying when there is an error or getHomeAz api is not supported by nmagent
+			if populateErr != nil || !homeAzResponse.IsSupported {
+				log.Debugf("Failed to retrieve home az from nmagent, will retry. %v", populateErr)
+				time.Sleep(time.Duration(retryIntervalInSecs) * time.Second)
```

Comment:
I'd prefer this to use a `time.Ticker`, so that `c.closing` has a chance to run outside of the sleep interval.